thriftbp/server.go

```diff
@@ -91,38 +97,201 @@ func NewServer(cfg ServerConfig) (*thrift.TSimpleServer, error) {
+
+		if err := WithDefaultMiddleware(bp)(cfg); err != nil {
+			return err
+		}
+
+		// Prepend defaults to existing values
+		cfg.Middlewares = append(cfg.Middlewares, existingMiddleware...)
+		return nil
+	}
+}
+
+// WithDefaultMiddleware will append the default baseplate middleware to the
+// server config. When depending on specific features such as ErrorSpanSuppressor,
+// the ServerOption configuring the setting needs to be invoked before this method.
+func WithDefaultMiddleware(bp baseplate.Baseplate) ServerOption {
+	return func(cfg *ServerConfig) error {
+		return WithMiddleware(BaseplateDefaultProcessorMiddlewares(
+			DefaultProcessorMiddlewaresArgs{
+				EdgeContextImpl:                    bp.EdgeContextImpl(),
+				ErrorSpanSuppressor:                cfg.ErrorSpanSuppressor,
+				ReportPayloadSizeMetricsSampleRate: cfg.ReportPayloadSizeMetricsSampleRate,
+			},
+		)...)(cfg)
+	}
+}
+
+func withDefaultLogger(bp baseplate.Baseplate) ServerOption {
+	return WithDefaultLogger(bp.GetConfig().Log.Level)
+}
+
+// withDefaultListenerAddress is used to copy the address config from the
+// baseplate config to the server config. Currently, it's only
+// needed to guarantee backwards compatibility in NewServer
+func withDefaultListenerAddress(bp baseplate.Baseplate) ServerOption {
+	return WithListenerAddress(bp.GetConfig().Addr)
+}
+
+// WithMiddleware appends the arg middleware to the end of
+// the cfg.Middleware slice
+func WithMiddleware(middleware ...thrift.ProcessorMiddleware) ServerOption {
+	return func(cfg *ServerConfig) error {
+		cfg.Middlewares = append(cfg.Middlewares, middleware...)
+		return nil
+	}
+}
+
+// WithLogger configures the thrift.Logger that will be used by the server
+// to log connection errors.
+func WithLogger(logger thrift.Logger) ServerOption {
+	return func(cfg *ServerConfig) error {
+		cfg.Logger = logger
+		return nil
+	}
+}
+
+// WithDefaultLogger initializes the logger that will be used by the thrift
+// server to log connection errors. The logger will be configured with the
+// level supplied.
+func WithDefaultLogger(level log.Level) ServerOption {
+	return WithLogger(
+		log.ZapWrapper(
+			log.ZapWrapperArgs{
+				Level: level,
+				KVPairs: map[string]interface{}{
+					"from": "thrift",
+				},
+			},
+		).ToThriftLogger(),
+	)
+}
+
+// WithListenerAddress is defined to provide backwards compatibility with
+// existing usages of NewServer. Use WithDefaultSocket instead.
+func WithListenerAddress(address string) ServerOption {
+	return func(cfg *ServerConfig) error {
+		cfg.Addr = address
+		return nil
+	}
+}
+
+// WithSocket creates a server option that initializes a
+// thrift.TServerSocket listening on the address supplied.
+func WithSocket(socket *thrift.TServerSocket) ServerOption {
+	return func(cfg *ServerConfig) error {
+		cfg.Socket = socket
+		return nil
+	}
+}
+
+// WithDefaultSocket initializes a socket listening at the
+// configured address.
+func WithDefaultSocket() ServerOption {
```

Comment:
what's the use-case for users providing this option themselves?

Either they have a ServerConfig (in which case they should be using WithConfigFrom) or they have an address (and should use WithListenerAddress) or socket (WithSocket)